internal/recommendations/client.go

```diff
@@ -0,0 +1,173 @@
+package recommendations
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+)
+
+type RecommendationsClient struct {
+	recommendationsURL string
+	tokenURL           string
+	accessToken        string
+	clientId           string
+	clientSecret       string
+	client             *http.Client
+	tokenMu            sync.RWMutex
+}
+
+type RecommendationsClientConfig struct {
+	RecommendationsURL string
+	CA                 string
+	TokenURL           string
+	ClientId           string
+	ClientSecret       string
+}
+
+type tokenResponse struct {
+	AccessToken string `json:"access_token"`
+}
+
+var contentHeaders = map[string]string{"Content-Type": "application/json"}
+
+func NewClient(conf RecommendationsClientConfig) (*RecommendationsClient, error) {
+	if conf.TokenURL == "" {
+		return nil, fmt.Errorf("client needs token endpoint")
+	}
+	if conf.ClientId == "" {
+		return nil, fmt.Errorf("client needs clientId")
+	}
+	if conf.ClientSecret == "" {
+		return nil, fmt.Errorf("client needs client secret")
+	}
+	client, err := createClient(conf.RecommendationsURL, conf.CA)
+	if err != nil {
+		return nil, fmt.Errorf("error creating recommend http client")
+	}
+
+	rc := RecommendationsClient{
+		recommendationsURL: conf.RecommendationsURL,
+		tokenURL:           conf.TokenURL,
+		clientId:           conf.ClientId,
+		clientSecret:       conf.ClientSecret,
+		client:             client,
+	}
+
+	return &rc, nil
+}
+
+func createClient(recommendationsURL string, ca string) (*http.Client, error) {
+	if !strings.HasPrefix(recommendationsURL, "https") || ca == "" {
+		return &http.Client{}, nil
+	}
+
+	var tlsConfig *tls.Config
+	caCert, err := os.ReadFile(filepath.Clean(ca))
+	if err != nil {
+		return nil, err
+	}
+
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(caCert)
+
+	tlsConfig = &tls.Config{
+		MinVersion: tls.VersionTLS12,
+		RootCAs:    caCertPool,
+	}
+
+	transport := &http.Transport{TLSClientConfig: tlsConfig}
+	return &http.Client{Transport: transport}, nil
+}
+
+func (rc *RecommendationsClient) request(method, url string, headers map[string]string, body io.ReadSeeker) (*http.Response, error) {
+	req, err := http.NewRequest(method, url, body)
+	if err != nil {
+		return nil, err
+	}
+
+	for k, v := range headers {
+		req.Header.Add(k, v)
+	}
+
+	token := func() string {
+		rc.tokenMu.RLock()
+		defer rc.tokenMu.RUnlock()
+		return rc.accessToken
+	}
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token()))
+
+	resp, err := rc.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode == http.StatusForbidden {
+		err = rc.refreshToken()
+		if err != nil {
+			return nil, err
+		}
+
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token()))
+		resp, err = rc.client.Do(req)
```

Comment:
Before redoing the request you need to reseek:
```suggestion
                _, err = body.Seek(0, io.SeekStart)
                if err != nil {
                        return nil, err
                }
		resp, err = rc.client.Do(req)
```

Reply:
amazing, its work!! thank you so much